Add string representation for AttrsDescriptor

alexbaden opened the pull request:
The string representation allows PyTorch Inductor to serialize/derserialize the `AttrsDescriptor` to the `@triton.heuristics` block in the generated code.

alexbaden commented:
@giuseros are you ok with these changes?

giuseros commented:
Hi @alexbaden . I am fine with it, but you might want to update this, because[ the dictionary became a bit more complex](https://github.com/triton-lang/triton/pull/4910). Also, you should add a test to make sure that further changes are reflected in the `__repr__` method.

jataylo commented:
@alexbaden @giuseros 

I was testing on latest inductor and hitting some issues can we make the `__repr__` more generic with something like

> def __repr__(self):
    return f"AttrsDescriptor.from_dict({self.to_dict()})"
    
 Note we'll also need a PR to patch https://github.com/pytorch/pytorch/blob/main/torch/_inductor/runtime/hints.py#L43 with the new dictionary.

alexbaden commented:
Thanks @jataylo - I took your recommendation and rebased and modified this patch accordingly. 

@giuseros I think this should resolve concerns around structural changes in `AttrsDescriptor` being represented in `repr` - as long as `toDict` and `fromDict` serialization works, `repr` will reflect the latest obj structure.

peterbell10 commented on python/triton/backends/compiler.py:
```diff
@@ -210,6 +210,9 @@ def get_property_key(val, align):
         return "N"
 
+    def __repr__(self):
+        return f"AttrsDescriptor.from_dict({self.to_dict()})"
```
Don't think it makes a difference currently, but this should probably format the dict's`repr`
```suggestion
        return f"AttrsDescriptor.from_dict({self.to_dict()!r})"
```

alexbaden replied:
Makes sense. Done.
